components/camel-kafka/src/main/java/org/apache/camel/processor/idempotent/kafka/KafkaIdempotentRepository.java

```diff
@@ -0,0 +1,342 @@
+                log.info("Cache OK");
+            } else {
+                log.warn("Timeout waiting for cache warm-up from topic {}. Proceeding anyway. "
+                        + "Duplicate records may not be detected.", topic);
+            }
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+
+    }
+
+    @Override
+    public CamelContext getCamelContext() {
+        return this.camelContext;
+    }
+
+    @Override
+    protected void doStop() {
+        // stop the thread
+        topicPoller.setRunning(false);
+        try {
+            if (topicPoller.getShutdownLatch().await(30, TimeUnit.SECONDS)) {
+                log.info("Expired waiting on topicPoller to shut down");
+            }
+        } catch (InterruptedException e) {
+            log.info("Interrupted waiting on latch: {}", e.getMessage());
+        }
+        executorService.shutdown();
+
+        try {
+            consumer.close();
+        } finally {
+            producer.close();
+        }
+    }
+
+    @Override
+    public boolean add(String key) {
+        if (cache.containsKey(key)) {
+            duplicateCount.incrementAndGet();
+            return false;
+        } else {
+            broadcastAction(key, CacheAction.add);
+            return true;
+        }
+    }
+
+    private void broadcastAction(String key, CacheAction action) {
+        try {
+            log.debug("Broadcasting action:{} for key:{}", action, key);
+            producer.send(new ProducerRecord<>(topic, key, action.toString())).get(); // sync send
+        } catch (ExecutionException | InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    @ManagedOperation(description = "Does the store contain the given key")
+    public boolean contains(String key) {
+        log.debug("Checking cache for key:{}", key);
+        boolean containsKey = cache.containsKey(key);
+        if (containsKey) {
+            duplicateCount.incrementAndGet();
+        }
+        return containsKey;
+    }
+
+    @Override
+    @ManagedOperation(description = "Remove the key from the store")
+    public boolean remove(String key) {
+        broadcastAction(key, CacheAction.remove);
+        return true;
+    }
+
+    @Override
+    public boolean confirm(String key) {
+        return true; // no-op
+    }
+
+    @Override
+    public void clear() {
+        broadcastAction(null, CacheAction.clear);
+    }
+
+    @ManagedOperation(description = "Number of times duplicate messages have been detected")
+    public long getDuplicateCount() {
+        return duplicateCount.get();
+    }
+
+    @ManagedOperation(description = "Number of times duplicate messages have been detected")
+    public boolean isPollerRunning() {
+        return topicPoller.getRunning();
+    }
+
+    private class TopicPoller implements Runnable {
+
+        private static final int POLL_DURATION_MS = 10;
```

Comment:
Is this poll to short? Maybe users want to be able to configure this. Isnt this very chatty over network to do 100 polls per second